File: CS333_Project_05/LinkedList.c
/*
Jordan Smith
LinkedList.c
10/25/23
Linked List
*/

#include <stdio.h>
#include <stdlib.h>

// Node struct to hold data and a pointer to the next Node
struct Node {
    void *data;           // Pointer to the stored data
    struct Node *next;    // Pointer to the next Node in the list
};

// LinkedList struct to hold the head of the list
struct LinkedList {
    struct Node *head;     // Pointer to the first Node in the list
    int size;              // Size of the list
};

// Function to create a new LinkedList
struct LinkedList *ll_create() {
    struct LinkedList *list = (struct LinkedList *)malloc(sizeof(struct LinkedList));
    if (list) {
        list->head = NULL;
        list->size = 0;
    }
    return list;
}

// Function to insert a new Node with data at the front of the list
void ll_push(struct LinkedList *list, void *data) {
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    if (newNode) {
        newNode->data = data;
        newNode->next = list->head;
        list->head = newNode;
        list->size++;
    }
}

// Function to remove and return the data from the front of the list
void *ll_pop(struct LinkedList *list) {
    if (list->head) {
        struct Node *temp = list->head;
        list->head = list->head->next;
        void *data = temp->data;
        free(temp);
        list->size--;
        return data;
    }
    return NULL;
}

// Function to add a new Node with data at the end of the list
void ll_append(struct LinkedList *list, void *data) {
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    if (newNode) {
        newNode->data = data;
        newNode->next = NULL;

        if (list->head == NULL) {
            list->head = newNode;
        } else {
            struct Node *current = list->head;
            while (current->next != NULL) {
                current = current->next;
            }
            current->next = newNode;
        }
        list->size++;
    }
}

// Function to remove the first node with data matching the target using the given comparison function
void *ll_remove(struct LinkedList *list, void *target, int (*compfunc)(void *, void *)) {
    struct Node *current = list->head;
    struct Node *prev = NULL;

    while (current != NULL) {
        if (compfunc(current->data, target) == 0) {
            if (prev != NULL) {
                prev->next = current->next;
            } else {
                list->head = current->next;
            }

            void *data = current->data;
            free(current);
            list->size--;
            return data;
        }

        prev = current;
        current = current->next;
    }

    return NULL;
}

// Function to get the size of the list
int ll_size(struct LinkedList *list) {
    return list->size;
}

// Function to clear the list and free associated data using the given function
void ll_clear(struct LinkedList *list, void (*freefunc)(void *)) {
    struct Node *current = list->head;
    while (current != NULL) {
        struct Node *temp = current;
        current = current->next;
        if (freefunc) {
            freefunc(temp->data);
        }
        free(temp);
    }
    list->head = NULL;
    list->size = 0;
}

// Function to apply a function to the data at each node
void ll_map(struct LinkedList *list, void (*mapfunc)(void *)) {
    struct Node *current = list->head;
    while (current != NULL) {
        mapfunc(current->data);
        current = current->next;
    }
}

File: CS333_Project_05/linkedlist.js
/*
Jordan Smith
linkedlist.js
10/25/23
Linked List
*/

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    // Creates a new LinkedList and returns it
    static ll_create() {
        return new LinkedList();
    }

    // Adds a node to the front of the list, storing the given data in the node
    ll_push(data) {
        const newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    // Removes the node at the front of the list and returns the associated data
    ll_pop() {
        if (!this.head) {
            return null;
        }
        const data = this.head.data;
        this.head = this.head.next;
        this.size--;
        return data;
    }

    // Adds a node to the end of the list, storing the given data in the node
    ll_append(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
        this.size++;
    }

    // Removes the first node in the list whose data matches the target using the comparison function
    // Returns the pointer to the data in the removed node
    ll_remove(target, compfunc) {
        if (!this.head) {
            return null;
        }
        if (compfunc(target, this.head.data)) {
            const data = this.head.data;
            this.head = this.head.next;
            this.size--;
            return data;
        }

        let current = this.head;
        while (current.next) {
            if (compfunc(target, current.next.data)) {
                const data = current.next.data;
                current.next = current.next.next;
                this.size--;
                return data;
            }
            current = current.next;
        }

        return null;
    }

    // Returns the size of the list
    ll_size() {
        return this.size;
    }

    // Removes all of the nodes from the list, freeing the associated data using the given function
    ll_clear(freefunc) {
        let current = this.head;
        while (current) {
            const nextNode = current.next;
            if (freefunc) {
                freefunc(current.data);
            }
            current = nextNode;
        }
        this.head = null;
        this.size = 0;
    }

    // Traverses the list and applies the given function to the data at each node
    ll_map(mapfunc) {
        let current = this.head;
        while (current) {
            if(mapfunc(current.data) != null)
            {
                current.data = mapfunc(current.data); // Update the data
            }
            current = current.next;
        }
    }    
}

module.exports = {LinkedList};